benches/nixpkgs/lists.ncl

```diff
@@ -116,4 +116,557 @@
        >> [ "a", "z", "b", "z" ]
   "%m
   = (fun f list => array.flatten (array.map f list)),
+
+  flatten: Dyn -> Array Dyn
+  | doc m%"
+      Flatten the argument into a single list, that is, nested lists are
+     spliced into the top-level lists.
+
+     Example:
+       flatten [1, [2, [3], 4], 5]
+       >> [1, 2, 3, 4, 5]
+       flatten 1
+       >> [1]
+  "%m
+  = fun x =>
+    if builtin.is_array x
+    then concatMap (fun y => flatten y) (x | Array Dyn)
+    else [x],
+
+  remove: forall a. a -> Array a -> Array a
+  | doc m%"
+      Remove elements equal to 'e' from a list.  Useful for buildInputs.
+
+     Example:
+       remove 3 [ 1, 3, 4, 3 ]
+       >> [ 1, 4 ]
+  "%m
+  = # Element to remove from the list
+    fun e => array.filter ((!=) e),
+
+  findSingle: forall a. (a -> Bool) -> a -> a -> Array a -> a
+  | doc m%"
+      Find the sole element in the list matching the specified
+     predicate, returns `default` if no such element exists, or
+     `multiple` if there are multiple matching elements.
+
+     Example:
+       findSingle (fun x => x == 3) "none" "multiple" [ 1, 3, 3 ]
```

Comment:
(beginner question)
Don't you have an issue with the types for this example?
The function type says that `def`, `multiple` & the elements of the array should have the same type, but in your example the elements of the array are integers and the others are string.

Or am I missing something?

Reply:
@bew Yes, you're right. You could still use this function with `a = Dyn` indeed, but Nickel won't infer this kind of upper bound for you automatically, so you would need a bunch of contract or type annotation to make it accepted. The above type is wrong (I mean it is sound, but arbitrarily too restrictive), and the provided example probably doesn't typecheck as it is.

I guess a more reasonable type is `forall a b. (a -> Bool) -> b -> b -> Array a -> b`

Reply:
Well no, because if the element is found only once, it should be returned..
And `def` & `multiple` could (for whatever reason) be of different types..

Is there a plan to add union types to Nickel and allow this to precisely return a kind of `Union[ElemType, NoneType, MultiType]` ?

Reply:
For some reasons I wrote an answer but probably forgot to hit "Enter"... @bew you're right, again, I just inferred a type from the example too quickly. Indeed, in a static type system I don't think we can do much better than the current type. As you say, the example is morally relying on union types.

Note that the example is fine if it is in dynamically typed code: the contract won't complain. In typed code, you can use some type annotations to make the example pass. Sorry, I've written a working example in the previous ghost comment, but I can't find it. It merely amounted to use a `: Dyn` for the result of the call and `| Dyn` and `| Array Dyn` for arguments. The boolean function argument didn't need an annotation, so that's not too bad. The RFC on typechecking could even help remove some of the `Dyn` annotations on the arguments.

Union types are not planned in the near future. We actually tried to implement them, but encountered difficulties: in Nickel, each static type must have a contract as a counter-part (for so-called *soundness* of the gradual typing), but it turns out combining generic unions with lazy contracts is really hard, with undesirable consequences. We even presented [a paper about this misadventure](https://arxiv.org/abs/2106.06278) at SPLASH21 and ECOOP22, which is condensed in [a blog post](https://www.tweag.io/blog/2022-04-28-union-intersection-contracts/).